p2p/discovery/table.go

```diff
@@ -24,51 +26,108 @@ type Table struct {
 	selfNode *Node //info of local node
 }
 
-type bucket struct {
-	peers []*Node
-}
-
 func NewTable(id NodeID, addr *net.UDPAddr) *Table {
-	selfNode := NewNode(id, addr)
+	selfNode := NewNodeWithAddr(id, addr)
 
 	table := &Table{
 		count:    0,
 		selfNode: selfNode,
 	}
 
+	for i := 0; i < nBuckets; i++ {
+		table.buckets[i] = NewBuckets()
+	}
+
 	return table
 }
 
-func (t *Table) AddNode(node *Node) {
+func (t *Table) addNode(node *Node) {
 	dis := logdist(node.sha, t.selfNode.sha)
 
-	t.buckets[dis].AddNode(node)
+	t.buckets[dis].addNode(node)
 }
 
-// AddNode add node to bucket, if bucket is full, will remove an old one
-func (b *bucket) AddNode(node *Node) {
-	index := b.HasNode(node)
+func (t *Table) updateNode(node *Node)  {
+	t.addNode(node)
+}
 
-	if index != -1 {
-		// do nothing for now
-		// TODO lru
-	} else {
-		if len(b.peers) < bucketSize {
-			b.peers = append(b.peers, node)
-		} else {
-			copy(b.peers[:], b.peers[1:])
-			b.peers[len(b.peers)-1] = node
+func (t *Table) findNodeResponse(target common.Hash) []*Node {
+	nodes := t.findMinDisNodes(target, responseNodeNumber)
+
+	minDis := []*Node{}
+	for _, e := range nodes {
+		if distcmp(target, t.selfNode.sha, e.sha) > 0 {
+			minDis = append(minDis, e)
 		}
 	}
+
+	return minDis
+}
+
+func (t *Table) deleteNode(target common.Hash)  {
```

Comment:
do we need to add lock/mutex for this func?

Reply:
yes. I will add it later